### DISCUSSION TITLE
Trouble Harmonizing Data!

### DISCUSSION THREAD

**Discussion by JulesMRuef** [earthlab-education/Earth-Analytics-2023-01-Intro#674] [! GitHub help] [answered]
Hello All! 
I'm running into quite a bit of trouble trying to harmonize my layers! I have a sneaking suspicion this might have something to do with the fact that my climate variable (precipitation in 2005 and 1980) is stored as a netcdf file as opposed to a tif file. I keep getting an error that reads "MissingCRS: CRS not found. Please set the CRS with 'rio.write_crs()'. Data variable: precipitation". Has anyone else dealt with this? 
Thanks in advance!

**Comment by advythr** [comment]
Hi Juliana,

I am not working with this same dataset but I had this same annoying error message. For me, ChatGPT was leading me astray.

Assuming the raster data is being stored as a rioxarray data array, I set the CRS with:
```
# Add in a CRS
data_array = data_array.rio.write_crs(rxr.crs.CRS.from_epsg(3857))

# Test if the CRS was added
print(data_array.rio.crs)
```

This code worked to resolve that error and after setting the CRS, I was able to plot the raster.

**Reply by NobodyLivesHere** [reply]
Yeah if you pull the variable like precipitation out of the dataset it will need to be reprojected and given spatial dimensions otherwise it will error out.  Other people were having this issue and iirc was covered in the video posted last week.

**Reply by peterkobylarz** [reply]
I'm running into the same issue with the error saying the 'precipitation' data variable has no coordinate system. However, my data array has been optimized for analysis and no longer has any attributes:
![image](https://github.com/earthlab-education/Earth-Analytics-2023-01-Intro/assets/143359014/76d68cac-eb05-45b5-b0f5-64607569caea)
I'm still getting the following error stating that the 'data variable' 'precipitation' has no coordinate system:
![image](https://github.com/earthlab-education/Earth-Analytics-2023-01-Intro/assets/143359014/6124c4ce-4a01-4ab9-a5d6-02d662a20c43)

My code is as follows:
```
harmonizer = data_accumulator['harmonizer']
harmonizer.rio.write_crs(rxr.crs.CRS.from_epsg(3857))
harmonize_test = precip_2055_optimal_indiangrass.rio.reproject_match(harmonizer)
```

Calling 'precip_2055_optimal.precipitation' returns an error saying no attribute exists... am I missing something here? How/where do I set the CRS for precipitation?

**Reply by peterkobylarz** [reply]
I tried using the harmonize process on a data array raster that hadn't been converted to boolean values for my analysis yet ant it worked. Here's the code I ended up using for reference:
```python
precip_2055_clip.rio.write_crs("epsg:3857", inplace=True)
print(precip_2055_clip.rio.crs)
harmonizer = data_accumulator['harmonizer']
harmonizer.rio.write_crs("epsg:3857", inplace=True)
print(harmonizer.rio.crs)
harmonize_test = precip_2055_clip.rio.reproject_match(harmonizer)
```

**Comment by misterskye** [comment, marked as answer]
Hey everyone---  After much deliberation and angst, I reached out to Elsa on trying to get the Precip data in alignment with the other rasters, and it worked!!  
Here is the code where we first set up the Precip data which has it in Lat/Lon:  

maca_ds = xr.open_dataset('maca.nc')
maca_ds = maca_ds.assign_coords(lon=maca_ds.lon-360)
precip_da = maca_ds.precipitation
precip_da.rio.write_crs('epsg:4326', inplace=True)
precip_da.rio.set_spatial_dims('lon', 'lat', inplace=True)

At this point, I wanted all of my rasters to match the Polaris Soil raster (ph_da), and so I had to not only reproject, but rename the dims.

precip_raster.rename(lon='x', lat = 'y').rio.reproject_match(soil_raster)

Hope this helps
